### PR TITLE
Use a whitelist for peers participating on consensus

### LINKED ISSUE #756: Consensus should run with whitelisted VPs when security is enabled
This is a follow-up to #462.

We currently require that VPs are named as `vp0`, `vp1`, ...`vpX` where `X+1` is the maximum number of validating nodes that we have set in the config file.  [As noted in #462](https://github.com/openblockchain/obc-peer/issues/462#issuecomment-173421403), this approach is flaky and needs to go.

The proper way to do this is: create a whitelist of hashed raw enrollment certificates that will be allowed to connect to the network. We dropped this because it was causing issues when automatically deploying a network through Swarm.

@adecaro: Can we modify the CA so that it pre-generates a list of N enrollment certificates. We could then use them to populate the network's whitelist. Later on, as each VP signs up, we'll have the CA return one of these pre-generated certificates. Are there any security implications we need to be aware of in this approach?

Another way to do this is: Wait for the first N validating nodes to connect to the network, and don't kick off consensus until this happens. These are your validators, this is your whitelist. (Chet is thinking along these lines as well IIRC.)

### PR DESCRIPTION
**Pull request by tuand27613**
## Description

Create a list of peers that can participate in consensus. The list is build from CORE_PBFT_GENERAL_N. Consensus does not start until N peers have connected. After that, only peers in the whitelist can reconnect.
The restriction on peers having to be named vp0/1/2/3,etc ... is also lifted.
## Motivation and Context

Fixes #756
## How Has This Been Tested?

Unit tests run ( unit tests were modified to use the whitelist methods )
Behave tests run ( the compose-consensus-4 docker-compose files in bddtests were modified so that peer names are not vp0/1/2/3 )
## Checklist:
- [X] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff)
- [X] Either no new documentation is required by this change, OR I added new documentation
- [X] Either no new tests are required by this change, OR I added new tests
- [X] I have run [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports), [go vet](https://golang.org/cmd/vet/), and [golint](https://github.com/golang/lint). I have cleaned up all valid errors and warnings in code I have added or modified. These tools may generate false positives. Don't be worried about ignoring some errors or warnings. The goal is clean, consistent, and readable code.

Signed-off-by: Tuan Dang tdang@us.ibm.com

### CODE REVIEW

**Comment by tuand27613** [comment]
@jyellick @corecode  @kchristidis please take a look.

File: core/peer/statetransfer/statetransfer.go [hunk 1]
```diff
@@ -25,11 +25,11 @@ import (
 	"time"
 
 	_ "github.com/hyperledger/fabric/core" // Logging format init
+	"github.com/op/go-logging"
```
Comment (corecode) [inline core/peer/statetransfer/statetransfer.go:28]: why?

File: core/peer/whitelist.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+	"strings"
+
+	"github.com/hyperledger/fabric/core/db"
+	pb "github.com/hyperledger/fabric/protos"
+	"github.com/spf13/viper"
+
+	"github.com/golang/protobuf/proto"
+)
+
+const whitelistFile = "/tmp/whitelist.dat"
```
Comment (corecode) [inline core/peer/whitelist.go:34]: not needed

File: core/peer/whitelist.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+// CheckWhitelistExists blocks until we have N validators in the whitelist
+func (p *PeerImpl) CheckWhitelistExists() {
+	<-p.whitelistCreated
+}
+
+// LoadWhitelist loads the validator's whitelist of validating peers from disk into memory
+// Sets p.whitelist and p.whitelistedMap
+func (p *PeerImpl) LoadWhitelist() (err error) {
+
```
Comment (corecode) [inline core/peer/whitelist.go:56]: empty line

File: core/peer/whitelist.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+	if !p.isValidator {
+		return nil
+	} // if I'm an NVP, I don't care
+
+	// pbft plugin waits on this channel for whitelist to complete
+	p.whitelistCreated = make(chan struct{}, 3) // make non-blocking. We only need to wake up the receiver once
```
Comment (corecode) [inline core/peer/whitelist.go:62]: why 3?

File: core/peer/whitelist.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+	}
+	peerLogger.Debug("Whitelist now reads: %+v", p.whitelist)
+
+	return err
+}
+
+// SaveWhitelist stores to disk the list of validating peers this validator should connect to
+// Is called when the handlerMap is locked by RegisterHandler
+func (p *PeerImpl) SaveWhitelist() (err error) {
+
```
Comment (corecode) [inline core/peer/whitelist.go:95]: empty line

File: bddtests/docker-compose-4-consensus-sieve.yml [hunk 6]
```diff
@@ -8,7 +8,7 @@ vp0:
     file: docker-compose-4-consensus-base.yml
     service: vpSieve
   environment:
-    - CORE_PEER_ID=vp0
```
Comment (corecode) [inline bddtests/docker-compose-4-consensus-sieve.yml:11]: why do these names change?

File: consensus/consensus.go [hunk 7]
```diff
@@ -30,8 +30,18 @@ type Consenter interface {
 
 // Inquirer is used to retrieve info about the validating network
 type Inquirer interface {
-	GetNetworkInfo() (self *pb.PeerEndpoint, network []*pb.PeerEndpoint, err error)
-	GetNetworkHandles() (self *pb.PeerID, network []*pb.PeerID, err error)
+	GetOwnID() (id uint64)
```
Comment (corecode) [inline consensus/consensus.go:33]: What do all of these interfaces do?  I have a suspicion that we don't need that many.

File: consensus/obcpbft/obc-batch.go [hunk 8]
```diff
@@ -90,20 +92,37 @@ func newObcBatch(id uint64, config *viper.Viper, stack consensus.Stack) *obcBatc
...
+	go op.waitForID(config)
+
 	go op.main()
 	return op
 }
 
+func (op *obcBatch) waitForID(config *viper.Viper) {
+	var id uint64
+
+	op.stack.CheckWhitelistExists()
```
Comment (corecode) [inline consensus/obcpbft/obc-batch.go:111]: I presume this is blocking?  Not obvious from the name.

File: consensus/obcpbft/obc-batch.go [hunk 9]
```diff
@@ -90,20 +92,37 @@ func newObcBatch(id uint64, config *viper.Viper, stack consensus.Stack) *obcBatc
...
-	op.deduplicator = newDeduplicator()
-
 	// create non-running timer
 	op.batchTimer = time.NewTimer(100 * time.Hour) // XXX ugly
 	op.batchTimer.Stop()
 
 	op.idleChan = make(chan struct{})
 	op.viewChanged = make(chan struct{})
 
+	go op.waitForID(config)
```
Comment (corecode) [inline consensus/obcpbft/obc-batch.go:102]: why does this go here, and not at the very top of the constructor, and have the rest of the setup be performed later?

File: consensus/obcpbft/obc-batch.go [hunk 10]
```diff
@@ -90,20 +92,37 @@ func newObcBatch(id uint64, config *viper.Viper, stack consensus.Stack) *obcBatc
...
+	id = op.stack.GetOwnID()
+	logger.Debug("Replica ID = %v", id)
+
+	// instantiate pbft-core
+	op.pbft = newPbftCore(id, config, op)
+
+	op.complainer = newComplainer(op, op.pbft.requestTimeout, op.pbft.requestTimeout)
+	op.deduplicator = newDeduplicator()
+
+	op.isSufficientlyConnected <- true
```
Comment (corecode) [inline consensus/obcpbft/obc-batch.go:121]: i suggest `close(op.isSufficientlyConnected)`

File: consensus/obcpbft/obc-batch.go [hunk 11]
```diff
@@ -408,6 +418,11 @@ func (op *obcBatch) resubmitStaleRequest(c custodyInfo) {
 
 // allow the primary to send a batch when the timer expires
 func (op *obcBatch) main() {
+
```
Comment (corecode) [inline consensus/obcpbft/obc-batch.go:421]: if `op.isSufficientlyConnected` is closed, this reduces to an unconditional `<- op.isSufficientlyConnected`

File: core/peer/whitelist.go [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+	if p.whitelist.Cap == noWhitelistNeeded {
+		peerLogger.Info("whitelist will not be saved. whitelist.Cap = %v", p.whitelist.Cap)
+		return nil
+	}
+
+	vpMap := filterHandlers(p.handlerMap.m) // who are we connected to currently ?
+
+	if p.whitelist.Persisted {
+		peerLogger.Debug("Whitelist has already been saved. This is a restart.")
+		minNeeded := ((p.whitelist.Cap - 1) / 3) * 2 //2f
```
Comment (corecode) [inline core/peer/whitelist.go:110]: This has to be calculated by consensus and not by the peer.

File: core/util/utils.go [hunk 13]
```diff
@@ -88,3 +97,53 @@ func GenerateHashFromSignature(path string, ctor string, args []string) []byte {
 	hash := ComputeCryptoHash(b)
 	return hash
 }
+
+// LoadFromDisk loads a file from disk
+func LoadFromDisk(filename string) (data []byte, err error) {
```
Comment (corecode) [inline core/util/utils.go:102]: not used

**Comment by corecode** [comment]
I couldn't see any tests introduced for the whitelist functionality.

**Comment by corecode** [comment]
Overall this feels like a big hack (which it is), and it is quite intrusive as well.  I don't like the way it ties up consensus and the rest of the stack even more.

I think we need to take a different approach that does not need such deep changes, especially as essentially a convenience stop-gap measure.

If the stack exposes `PeerConnect()` and `PeerDisconnect()` events to consensus, most of the membership management can be taken care of in consensus itself.

**Comment by kchristidis** [comment]
I agree with @corecode's review. @tuand27613, I'll be happy to tackle this in a proper fashion, so if you want it off your hands let me know - don't want to get it in your way though.

**Comment by tuand27613** [comment]
I'll refactor on top of Jason's event manager work.  Thanks @kchristidis !

**Comment by ghost** [comment]
Can one of the admins verify this patch?

**Comment by tuand27613** [comment]
we're re-designing how we should handle whitelisting of peers via #756.

closing this PR.

**Pull request closed without merging**